lxd/instances_put.go

```diff
@@ -79,237 +77,37 @@ func containersPut(d *Daemon, r *http.Request) response.Response {
 	// Don't mess with containers while in setup mode
 	<-d.readyChan
 
-	// Batch the actions.
-	var opType db.OperationType
-	var do func(*operations.Operation) error
-	switch action {
-	case shared.Start:
-		opType = db.OperationInstanceStart
-		do = func(op *operations.Operation) error {
-			failures := map[string]error{}
-			failuresLock := sync.Mutex{}
-			wgAction := sync.WaitGroup{}
-
-			for _, inst := range instances {
-				wgAction.Add(1)
-				go func(inst instance.Instance) {
-					defer wgAction.Done()
-
-					err = inst.Start(raw.Stateful)
-					if err != nil {
-						failuresLock.Lock()
-						failures[inst.Name()] = err
-						failuresLock.Unlock()
-					}
-				}(inst)
-			}
-
-			wgAction.Wait()
-			return coalesceErrors(failures)
-		}
-	case shared.Stop:
-		opType = db.OperationInstanceStop
-		if raw.Stateful {
-			do = func(op *operations.Operation) error {
-				failures := map[string]error{}
-				failuresLock := sync.Mutex{}
-				wgAction := sync.WaitGroup{}
-
-				for _, inst := range instances {
-					wgAction.Add(1)
-					go func(inst instance.Instance) {
-						defer wgAction.Done()
-
-						err = inst.Stop(raw.Stateful)
-						if err != nil {
-							failuresLock.Lock()
-							failures[inst.Name()] = err
-							failuresLock.Unlock()
-						}
-					}(inst)
-				}
-
-				wgAction.Wait()
-				return coalesceErrors(failures)
-			}
-		} else if raw.Timeout == 0 || raw.Force {
-			do = func(op *operations.Operation) error {
-				failures := map[string]error{}
-				failuresLock := sync.Mutex{}
-				wgAction := sync.WaitGroup{}
-
-				for _, inst := range instances {
-					wgAction.Add(1)
-					go func(inst instance.Instance) {
-						defer wgAction.Done()
-
-						err = inst.Stop(false)
-						if err != nil {
-							failuresLock.Lock()
-							failures[inst.Name()] = err
-							failuresLock.Unlock()
-						}
-					}(inst)
-				}
-
-				wgAction.Wait()
-				return coalesceErrors(failures)
-			}
-		} else {
-			do = func(op *operations.Operation) error {
-				failures := map[string]error{}
-				failuresLock := sync.Mutex{}
-				wgAction := sync.WaitGroup{}
-
-				for _, inst := range instances {
-					wgAction.Add(1)
-					go func(inst instance.Instance) {
-						defer wgAction.Done()
-
-						if inst.IsFrozen() {
-							err := inst.Unfreeze()
-							if err != nil {
-								failuresLock.Lock()
-								failures[inst.Name()] = err
-								failuresLock.Unlock()
-								return
-							}
-						}
+	// Determine operation type.
+	opType, err := instanceActionToOptype(req.State.Action)
+	if err != nil {
+		return response.BadRequest(err)
+	}
 
-						err = inst.Shutdown(time.Duration(raw.Timeout) * time.Second)
-						if err != nil {
-							failuresLock.Lock()
-							failures[inst.Name()] = err
-							failuresLock.Unlock()
-						}
-					}(inst)
+	// Batch the changes.
```

Comment:
@stgraber these vars could be initialised inside the `do` function to limit scope, as far as I can see they are not used outside of it.

Reply:
done